### PR TITLE
Search box

### PR DESCRIPTION
**Pull request by spireken**

### CODE REVIEW

File: Applications/Spire/Source/Ui/SearchBox.cpp [hunk 1]
```diff
@@ -0,0 +1,81 @@
+#include "Spire/Ui/SearchBox.hpp"
+#include <QHBoxLayout>
+#include <QKeyEvent>
```
Comment (shaobinspire) [inline Applications/Spire/Source/Ui/SearchBox.cpp:3]: The header file is not used.

**Review by shaobinspire — CHANGES_REQUESTED** [review]

**Review by shaobinspire — APPROVED** [review]
Looks great.

**Comment by kranar** [comment]
Does this need more changes or is it good to go?

**Comment by spireken** [comment]
> Does this need more changes or is it good to go?

The last update that I got was Jon giving the okay, so as far as I know it just needs your review.

File: Applications/Spire/Include/Spire/Ui/SearchBox.hpp [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+  /** Displays a text search input box. */
+  class SearchBox : public QWidget {
+    public:
+
+      /**
+       * Constructs a SearchBox.
+       * @param parent The parent widget.
+       */
+      explicit SearchBox(QWidget* parent = nullptr);
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Ui/SearchBox.hpp:17]: Needs a model.

File: Applications/Spire/Include/Spire/Ui/SearchBox.hpp [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+#define SPIRE_SEARCH_BOX_HPP
+#include <QWidget>
+#include "Spire/Ui/TextBox.hpp"
+#include "Spire/Ui/Ui.hpp"
+
+namespace Spire {
+
+  /** Displays a text search input box. */
+  class SearchBox : public QWidget {
+    public:
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Ui/SearchBox.hpp:11]: Missing a way to indicate that the user is submitting the search.

File: Applications/Spire/Source/Ui/SearchBox.cpp [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  container_layout->addWidget(m_delete_button);
+  auto box = new Box(container, this);
+  auto layout = new QHBoxLayout(this);
+  layout->setContentsMargins({});
+  layout->addWidget(box);
+  proxy_style(*this, *box);
+  set_style(*this, DEFAULT_STYLE());
+  m_current_connection = m_text_box->get_model()->connect_current_signal(
+    [=] (const auto& current) {
+      m_delete_button->setVisible(!current.isEmpty());
```
Comment (kranar) [inline Applications/Spire/Source/Ui/SearchBox.cpp:71]: Move this logic out of the constructor.

File: Applications/Spire/Source/Ui/SearchBox.cpp [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    });
+  m_delete_button->connect_clicked_signal([=] {
+    m_text_box->get_model()->set_current({});
```
Comment (kranar) [inline Applications/Spire/Source/Ui/SearchBox.cpp:74]: Same here.

**Review by kranar — CHANGES_REQUESTED** [review]
Looks good.

**Review by kranar — APPROVED** [review]

**Pull request merged**